Debezium Embedded with Mariadb Connector - gtidSet NPE [DBZ-9243]

Migrated from [DBZ-9243](https://issues.redhat.com/browse/DBZ-9243)

h1. Bug report
h2. What Debezium connector do you use and what version?

io.debezium.connector.mariadb.MariaDbConnector:3.2.0.Final

debezium-connector-binlog:3.2.0.Final
h2. What is the connector configuration?

 
{code:java}
io.debezium.config.Configuration.create()
       .with("name", "mariadb-connector")
       .with("connector.class", "io.debezium.connector.mariadb.MariaDbConnector")
       .with("offset.storage",  "org.apache.kafka.connect.storage.FileOffsetBackingStore")
       .with("offset.storage.file.filename", "temp-folder/offset/offset.dat")
       .with("offset.flush.interval.ms", 60000)
       .with("schema.history.internal", "io.debezium.storage.file.history.FileSchemaHistory")
       .with("schema.history.internal.file.filename", "temp-folder/schema-history/schema.dat")
       .with("database.server.id", "85744")
       .with("database.hostname", "localhost")
       .with("database.port", 3306)
       .with("database.user", "root")
       .with("database.password", "root")
       .with("database.include.list", "trial")
       .with("table.include.list", "trial-2")
       .with("topic.prefix", "mariadb-connector")
       .build(){code}
 
h2. What is the captured database version and mode of deployment?

MariaDB@10.5 in my local PC running in a docker container
h2. What behavior do you see?

 
{code:java}
Caused by: java.lang.NullPointerException: Cannot invoke "com.github.shyiko.mysql.binlog.GtidSet.toString()" because "this.gtidSet" is null	at com.github.shyiko.mysql.binlog.BinaryLogClient.requestBinaryLogStreamMaria(BinaryLogClient.java:827) ~[mysql-binlog-connector-java-0.40.2.jar:0.40.2]	at com.github.shyiko.mysql.binlog.BinaryLogClient.requestBinaryLogStream(BinaryLogClient.java:793) ~[mysql-binlog-connector-java-0.40.2.jar:0.40.2]	at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:619) ~[mysql-binlog-connector-java-0.40.2.jar:0.40.2]	at com.github.shyiko.mysql.binlog.BinaryLogClient$7.run(BinaryLogClient.java:959) ~[mysql-binlog-connector-java-0.40.2.jar:0.40.2]{code}
binlog is ON mariadb side. Here is also gtid variables in db:
{code:java}
gtid_binlog_pos    0-1-1
gtid_binlog_state    0-1-1
gtid_cleanup_batch_size    64
gtid_current_pos    0-1-1
gtid_domain_id    0
gtid_ignore_duplicates    OFF
gtid_pos_auto_engines    
gtid_slave_pos    
gtid_strict_mode    OFF
wsrep_gtid_domain_id    0
wsrep_gtid_mode    OFF{code}
For whatever reason, I'm always getting this error saying that gtidSet is not set. Then I examined code and noticed that gtidSet is only set in here:

!image-2025-07-22-18-54-50-971.png!

Since I have no filteredGtidSet, client gtidSet is being set as empty string. But in BinaryLogClient:
!image-2025-07-22-18-56-06-713.png!

gtidSet is not being set if gtidStr is empty string which is the case I did mentioned. That concludes with gtidSet being null, and this is the part where the code crashes with NPE:
!image-2025-07-22-18-57-22-980.png!

I cannot disable gtid as it is harcoded unlike in mysql. How should I proceed? Are you aware of this, maybe I'm missing something else

 

 

--- a/debezium-connector-mariadb/src/test/java/io/debezium/connector/mariadb/MariaDbConnectorIT.java
+++ b/debezium-connector-mariadb/src/test/java/io/debezium/connector/mariadb/MariaDbConnectorIT.java
@@ -7,14 +7,30 @@ package io.debezium.connector.mariadb;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Map;
+import java.util.logging.Handler;
+import java.util.logging.Level;
+import java.util.logging.LogRecord;
+import java.util.logging.Logger;
 
 import org.apache.kafka.common.config.Config;
+import org.junit.jupiter.api.Test;
 
+import com.github.shyiko.mysql.binlog.BinaryLogClient;
+
+import io.debezium.config.CommonConnectorConfig;
 import io.debezium.config.Configuration;
 import io.debezium.config.Field;
+import io.debezium.connector.binlog.BinlogConnectorConfig;
+import io.debezium.connector.binlog.BinlogConnectorConfig.SnapshotMode;
 import io.debezium.connector.binlog.BinlogConnectorIT;
+import io.debezium.connector.binlog.util.BinlogTestConnection;
+import io.debezium.connector.binlog.util.UniqueDatabase;
 import io.debezium.connector.mariadb.MariaDbConnectorConfig.SnapshotLockingMode;
+import io.debezium.doc.FixFor;
+import io.debezium.jdbc.JdbcConnection;
 
 /**
  * @author Chris Cranford
@@ -75,5 +91,92 @@ public class MariaDbConnectorIT extends BinlogConnectorIT<MariaDbConnector, Mari
     protected String getExpectedQuery(String statement) {
 
         return "SET STATEMENT max_statement_time=600 FOR " + statement;
+    }
+
+    @Test
+    @FixFor("debezium/dbz#1378")
+    public void shouldStartWithEmptyGtidSet() throws Exception {
+        final UniqueDatabase database = getDatabase();
+
+        final Configuration initialConfig = database.defaultConfig()
+                .with(BinlogConnectorConfig.TABLE_INCLUDE_LIST, database.qualifiedTableName("products"))
+                .with(BinlogConnectorConfig.INCLUDE_SCHEMA_CHANGES, false)
+                .with(BinlogConnectorConfig.SNAPSHOT_MODE, SnapshotMode.NEVER)
+                .build();
+
+        start(getConnectorClass(), initialConfig);
+        waitForStreamingRunning(getConnectorName(), database.getServerName());
+
+        // Generate at least one GTID and persist it into the offset.
+        try (BinlogTestConnection db = getTestDatabaseConnection(database.getDatabaseName());
+                JdbcConnection connection = db.connect()) {
+            connection.execute("INSERT INTO products VALUES (default,'dbz-1378-prime','Prime',10.50)");
+        }
+
+        final SourceRecords firstRunRecords = consumeRecordsByTopic(1);
+        assertThat(firstRunRecords.recordsForTopic(database.topicForTable("products")).size()).isEqualTo(1);
+
+        stopConnector();
+
+        // Verify restart offset actually contains GTID state before applying excludes.
+        final String serverName = initialConfig.getString(CommonConnectorConfig.TOPIC_PREFIX);
+        final Map<String, String> partition = createPartition(serverName, database.getDatabaseName()).getSourcePartition();
+        final Map<String, ?> lastCommittedOffset = readLastCommittedOffset(initialConfig, partition);
+        final MariaDbOffsetContext offsetContext = loadOffsets(Configuration.create()
+                .with(CommonConnectorConfig.TOPIC_PREFIX, serverName)
+                .build(), lastCommittedOffset);
+        assertThat(offsetContext.gtidSet()).isNotBlank();
+
+        final Configuration restartConfig = database.defaultConfig()
+                .with(BinlogConnectorConfig.TABLE_INCLUDE_LIST, database.qualifiedTableName("products"))
+                .with(BinlogConnectorConfig.INCLUDE_SCHEMA_CHANGES, false)
+                .with(BinlogConnectorConfig.GTID_SOURCE_EXCLUDES, ".*")
+                .with(BinlogConnectorConfig.SNAPSHOT_MODE, SnapshotMode.NEVER)
+                .build();
+
+        final Logger binlogClientLogger = Logger.getLogger(BinaryLogClient.class.getName());
+        final List<String> secondStartMessages = new ArrayList<>();
+        final Handler captureHandler = new Handler() {
+            @Override
+            public void publish(LogRecord record) {
+                if (record != null && record.getLevel().intValue() >= Level.INFO.intValue()) {
+                    secondStartMessages.add(record.getMessage());
+                }
+            }
+
+            @Override
+            public void flush() {
+            }
+
+            @Override
+            public void close() {
+            }
+        };
+
+        binlogClientLogger.addHandler(captureHandler);
+        try {
+            // Re-start with GTID excludes so filtered GTID becomes empty and exercises the problematic path.
+            start(getConnectorClass(), restartConfig);
+            waitForStreamingRunning(getConnectorName(), database.getServerName());
+
+            try (BinlogTestConnection db = getTestDatabaseConnection(database.getDatabaseName());
+                    JdbcConnection connection = db.connect()) {
+                connection.execute("INSERT INTO products VALUES (default,'dbz-1378-restart','Restart',11.50)");
+            }
+
+            final SourceRecords secondRunRecords = consumeRecordsByTopic(1);
+            assertThat(secondRunRecords.recordsForTopic(database.topicForTable("products")).size()).isEqualTo(1);
+            assertConnectorIsRunning();
+        }
+        finally {
+            binlogClientLogger.removeHandler(captureHandler);
+        }
+
+        // With the upstream fix (0.40.7+), empty GTID state must not use MariaDB GTID connect-state path.
+        // Older behavior (0.40.6) logs this line and follows the buggy path.
+        assertThat(secondStartMessages)
+                .noneMatch(message -> message != null && message.startsWith("Requesting streaming from GTID set:"));
+
+        stopConnector();
     }
 }

--- a/pom.xml
+++ b/pom.xml
@@ -136,7 +136,7 @@
         <!-- Database drivers, should align with databases -->
         <version.postgresql.driver>42.7.7</version.postgresql.driver>
         <version.mysql.driver>9.1.0</version.mysql.driver>
-        <version.mysql.binlog>0.40.6</version.mysql.binlog>
+        <version.mysql.binlog>0.40.7</version.mysql.binlog>
         <version.mongo.driver>5.6.2</version.mongo.driver>
         <version.sqlserver.driver>12.4.2.jre8</version.sqlserver.driver>
         <version.db2.driver>11.5.0.0</version.db2.driver>

